iocore/net/quic/QUICCongestionController.cc

```diff
@@ -59,34 +60,65 @@ QUICCongestionController::on_packet_sent(size_t bytes_sent)
 }
 
 bool
-QUICCongestionController::_in_recovery(QUICPacketNumber packet_number)
+QUICCongestionController::_in_recovery(ink_hrtime sent_time)
 {
-  return packet_number <= this->_end_of_recovery;
+  return sent_time <= this->_recovery_start_time;
 }
 
 void
 QUICCongestionController::on_packet_acked(const PacketInfo &acked_packet)
 {
   // Remove from bytes_in_flight.
   SCOPED_MUTEX_LOCK(lock, this->_cc_mutex, this_ethread());
-  this->_bytes_in_flight -= acked_packet.bytes;
-  if (this->_in_recovery(acked_packet.packet_number)) {
+  this->_bytes_in_flight -= acked_packet.sent_bytes;
+  if (this->_in_recovery(acked_packet.time_sent)) {
     // Do not increase congestion window in recovery period.
     return;
   }
   if (this->_congestion_window < this->_ssthresh) {
     // Slow start.
-    this->_congestion_window += acked_packet.bytes;
+    this->_congestion_window += acked_packet.sent_bytes;
     QUICCCDebug("slow start window chaged");
   } else {
     // Congestion avoidance.
-    this->_congestion_window += this->_k_default_mss * acked_packet.bytes / this->_congestion_window;
+    this->_congestion_window += this->_k_max_datagram_size * acked_packet.sent_bytes / this->_congestion_window;
     QUICCCDebug("Congestion avoidance window changed");
   }
 }
 
 void
-QUICCongestionController::on_packets_lost(const std::map<QUICPacketNumber, PacketInfo *> &lost_packets)
+QUICCongestionController::_congestion_event(ink_hrtime sent_time, uint32_t pto_count)
```

Comment:
Different from specs because we need `pto_count`

```
      CongestionEvent(sent_time):
        // Start a new congestion event if the sent time is larger
        // than the start time of the previous recovery epoch.
        if (!InRecovery(sent_time)):
          recovery_start_time = Now()
          congestion_window *= kLossReductionFactor
          congestion_window = max(congestion_window, kMinimumWindow)
          ssthresh = congestion_window
          // Collapse congestion window if persistent congestion
          if (pto_count > kPersistentCongestionThreshold):
            congestion_window = kMinimumWindow
```

Reply:
> Different from specs because we need `pto_count`

It looks fine to me. I asked WG about dependency for LD from CC a few months ago because some variables used in CC are defined in Loss Detection section on the spec, and I suggested passing the variables as parameters for functions in CC, like you did. They understand the dependency issue, but the conclusion was like "leave it for implementations".

Can you add a comment about the difference so we can remember it's an intentional change.

Reply:
will do.